### PR TITLE
Multicluster support for kiali cache

### LINKED ISSUE #5724: Cache supports multiple kube backends
For multicluster scenarios, the cache should support multiple kubernetes backends. Aggregation of objects across the backends should be handled in the business layer but the kiali cache needs to manage those backends and provide them to the business layer. 

#5618

### PR DESCRIPTION
**Pull request by nrfox**
Adds support for multiple kube backends to the kiali cache. Splits out the kubernetes part of the cache into its own implementation. Keeps a single global kiali cache. Adds a cluster client manager to create/manage clients from different clusters.

Fixes #5724

### CODE REVIEW

File: kubernetes/cache/proxy_status.go
Comment (leandroberetta) [inline kubernetes/cache/proxy_status.go:43]: Just a thought, in some other places I read that eventually we will need to treat all clusters equally with no home cluster distinction but in some other places I see this method, I think in some way, we will need to define that one of them is the "home", right?. Not sure why yet but sounds logical.

**Review by leandroberetta — APPROVED** [review]
Nick, great work, looks good to me. I have just some comments but to me it's ready for merge.

File: business/workloads.go
Comment (nrfox) [reply business/workloads.go:57]: Yes whenever the workload service gets updated to support multicluster, then the individual k8s client can get removed from this service. I had included some of those changes with this PR originally but removed them so that this change would mostly be about updating the cache.

File: kubernetes/cache/proxy_status.go
Comment (nrfox) [reply kubernetes/cache/proxy_status.go:43]: Yeah there's a couple different services that do autodiscovery that will need to know the "home" cluster at least in addition to this case where you want to use the client of the cluster kiali is deployed in to do the polling of istiod.

**Comment by nrfox** [comment]
@leandroberetta I noticed that the cache was not being refreshed when the Kiali service account token changed which is a regression. I fixed that and added a few unit tests for that scenario if you could PTAL again.

File: kubernetes/cache/proxy_status.go
Comment (israel-hdez) [inline kubernetes/cache/proxy_status.go:43]: I think this is OK as we are starting with a single control plane. But as we add support to other models, this may need to evolve, as I think we won't be able to tie this proxy status cache to a cluster cache, nor the home cluster will be the only one having a control plane.

**Review by israel-hdez — COMMENTED** [review]
Just flushing some comments, as I think I need to give a closer look to the code.

BTW, I wonder if it would be better to merge this PR after the release? It looks a little risky to merge this week because of the refactoring.

File: business/workloads.go
Comment (nrfox) [reply business/workloads.go:129]: Whoops I must've removed this while refactoring by mistake. Good catch, I'll add it back in.

File: business/layer.go
Comment (nrfox) [reply business/layer.go:192]: Ideally it wouldn't be `"home"` but come from the istio config that Kiali is deployed alongside of. The other cluster names will come from the remote secret kubeconfig files. I'm unsure if these need to match with their name in their cluster's istio config but they at least need to be unique. At any rate, the `home` cluster is just the cluster that Kiali is deployed in. It's at least helpful while we are iteratively adding multi cluster support and I think it will be useful for some of the business services that gather info about the mesh and the cluster but it may not make sense to keep in the future. I'll add a global const for it.

File: kubernetes/cache/proxy_status.go
Comment (nrfox) [reply kubernetes/cache/proxy_status.go:43]: Yes we'll need to scrape all of them and most likely merge the results together.

**Comment by nrfox** [comment]
> BTW, I wonder if it would be better to merge this PR after the release? It looks a little risky to merge this week because of the refactoring.

Yes it'd probably be better to let this soak at the beginning of a release cycle given the large refactor. Someone else may need to merge this and carry the torch then if it will make it into the next release since I will most likely be unable to work on this during that time period.

**Comment by nrfox** [comment]
I removed the workload svc changes from this PR. This change now only updates the cache. The workload svc changes couldn't be tested yet without the ability to load multiple clusters in and those changes were more half baked. I originally included those as an example for how the different business services could be changed to support multicluster but I believe they'd be better served as their own PRs.

File: kubernetes/cache/cache.go
Comment (israel-hdez) [inline kubernetes/cache/cache.go:143]: So, only the "Home cluster" client is updated. Perhaps it isn't clear in this PR, but I guess this is because the bound SA token only applies to the "local" cluster, while for remote ones we will be using standard non-expiring tokens.

Nothing to fix, just pointing out this.

File: kubernetes/client_factory.go
Comment (israel-hdez) [inline kubernetes/client_factory.go:170]: I guess this `TODO` means that this `newSAClient` func isn't really capable of creating a client for remote clusters, right?.

If so, I guess that, although we can say that the cache would be ready for MC after merging this PR, the client factory is not ready yet.

File: kubernetes/client_factory.go
Comment (israel-hdez) [inline kubernetes/client_factory.go:313]: I think a `TODO` is missing somewhere around this func, no?

If I understand correctly, it tries to compare the token of a remote cluster client, with the Kiali SA token (which would be for the local cluster). That's always going to be a mismatch, no? So, it will always want to recreate the client.

(this is one of the reasons why I think `clientfactory` is not ready for MC yet... although I see you have a unit test for this, so it is confusing).

**Review by israel-hdez — COMMENTED** [review]
Just a few comments after my 2nd review round.

Most likely, nothing major. Probably, just questions that need an answer.

File: kubernetes/cache/cache.go
Comment (nrfox) [reply kubernetes/cache/cache.go:143]: Yeah for remote clusters, I don't think the token will expire and even if it did the auto-renew mechanism that exists for the in-cluster token wouldn't auto-renew the out of cluster tokens. That being said, it could be good to watch for client changes for the other clients anyways in case the secrets change. I think that's something we can add later though when more of the pieces have come together and we can test this in earnest and have a clearer picture of the requirements around that. For now, just watching the client changes for the home cluster keeps the existing functionality of token renewal in single cluster which is why I added this.

File: kubernetes/client_factory.go
Comment (nrfox) [reply kubernetes/client_factory.go:170]: That's right. The cache will start a kube cache for each kiali service account client from the client factory but the client factory only supports a single service account client for now. It will need to be updated to read in the remote secrets from the filesystem and create a kiali service account client for each remote secret. This can happen after #5752 is merged.

File: kubernetes/client_factory.go
Comment (nrfox) [reply kubernetes/client_factory.go:313]: Yeah at the very least it needs a `TODO`. It should not be comparing to the kiali token for each cluster. Rather it should be comparing the client's token with the token read from the filesystem for that cluster. The unit for the home cluster is valid but the one for multiple clients is a bit misleading since we won't want to test with kiali token but rather with the token for that cluster. I'll remove the multiple cluster test since that will change anyways once the client factory supports reading the secrets from the fs.

Comment (nrfox) [reply kubernetes/client_factory.go:313]: Well looking at the test again it's actually testing two different methods. One test is for `GetSAHomeClusterClient` and the other test is for `GetSAClients`. In both cases we want to test that you get an updated client back in return but I'll update the test for `GetSAClients` so it's more clear.

**Review by israel-hdez — APPROVED** [review]
OK, I _think_ that I understand all changes. So, I'm going to approve, although I haven't tried the code.

I'll try it tomorrow. If I find something wrong, I'll remove my approval. Otherwise, you already have my green tick mark :-)

**Review by leandroberetta — APPROVED** [review]
Second review, still looks good to me.

**Comment by matejnesuta** [comment]
Hello. Can this possibly be tested with the integration suite as well? I am trying to find spots where the coverage can be increased and this looks promising, but I am not 100% sure.

**Comment by jshaughn** [comment]
> Hello. Can this possibly be tested with the integration suite as well? I am trying to find spots where the coverage can be increased and this looks promising, but I am not 100% sure.

The cache use is accessed via almost all code-paths, so there should be good coverage given existing and future tests.  I'm not sure that anything needs to be added here explicitly, but I'll let @nrfox comment when he returns from leave.

**Pull request merged**
